README.md
# employee_database

Loads [`test_db`](https://github.com/datacharmer/test_db) into a `mysql` database.  App nodes determine query and display it via web server.

## Setup

Install [`chefdk`](https://downloads.chef.io/chefdk/) this can be done by the following ways:
* by installer found [on the website](https://downloads.chef.io/chefdk/)
* with [`brew`](https://brew.sh)
    * `brew cask install chefdk`

Install `Virtual Box` this, also, can be done by the following ways:
* by installer found [on the website](https://www.virtualbox.org/wiki/Downloads)
* with [`brew`](https://brew.sh)
    * `brew cask install virtualbox`

Install [`Vagrant`](https://www.vagrantup.com/) this be done:
* by installer found [on the website](https://www.vagrantup.com/downloads.html)
* with [`brew`](https://brew.sh)
    * `brew cask install vagrant`

Install [`kitchen-vagrant`]:
    * `chef gem install kitchen-vagrant`

## Utilizing `Test Kitchen`
## Running

1) Set local shell to use `chef_shell`:
    `eval "$(chef shell-init bash)"`

1) Run `Test Kitchen` setup with:
    `kitchen converge`

1) Wait for infrastructure to be built.

1) Once the infrastructure is built, view load balancer at [`192.168.0.15`](http://192.168.0.15)

### Testing

In order to run tests, run the following:
`kitchen verify <suite_name>`
_where <suite_name> is the name of the node_

For example:
`kitchen verify database`

### Clean up

Once finished with building and testing infrastructure, run the following command to remove everything:

`kitchen destroy`

CHANGELOG.md
# employee_database CHANGELOG

This file is used to list changes made in each version of the employee_database cookbook.

# 0.1.0

Initial release.

- change 0
- change 1



spec/unit/recipes/default_spec.rb
#
# Cookbook:: employee_database
# Spec:: default
#
# Copyright:: 2019, The Authors, All Rights Reserved.

require 'spec_helper'

describe 'employee_database::default' do
  context 'When all attributes are default, on Ubuntu 16.04' do
    let(:chef_run) do
      # for a complete list of available platforms and versions see:
      # https://github.com/customink/fauxhai/blob/master/PLATFORMS.md
      runner = ChefSpec::ServerRunner.new(platform: 'ubuntu', version: '16.04')
      runner.converge(described_recipe)
    end

    it 'converges successfully' do
      expect { chef_run }.to_not raise_error
    end
  end

  context 'When all attributes are default, on CentOS 7.4.1708' do
    let(:chef_run) do
      # for a complete list of available platforms and versions see:
      # https://github.com/customink/fauxhai/blob/master/PLATFORMS.md
      runner = ChefSpec::ServerRunner.new(platform: 'centos', version: '7.4.1708')
      runner.converge(described_recipe)
    end

    it 'converges successfully' do
      expect { chef_run }.to_not raise_error
    end
  end
end


test/integration/default/lb_test.rb
# # encoding: utf-8

# Inspec test for recipe employee_database::lb

# The Inspec reference, with examples and extensive documentation, can be
# found at http://inspec.io/docs/reference/resources/

describe package('nginx') do
  it { should be_installed }
end

describe file('/etc/nginx/conf.d/default.conf') do
  it { should_not exist }
end

describe file('/etc/nginx/sites-enabled/000-default') do
  it { should_not exist }
end

describe file('/etc/nginx/conf.d/load-balancer.conf') do
  it { should exist }
  its('type') { should cmp 'file' }
  it { should be_file }
  it { should_not be_directory }
  its('size') { should be > 512 }
  its('mode') { should cmp '00644' }
end

describe port(80) do
  it { should be_listening }
end

describe port(443) do
  it { should_not be_listening }
end

test/integration/default/app_test.rb
# # encoding: utf-8

# Inspec test for recipe employee_database::lb

# The Inspec reference, with examples and extensive documentation, can be
# found at http://inspec.io/docs/reference/resources/

describe package('nginx') do
  it { should be_installed }
end

describe file('/usr/share/nginx/html/query_output.txt') do
  it { should exist }
  its('type') { should cmp 'file' }
  it { should be_file }
  it { should_not be_directory }
end

describe file('/usr/share/nginx/html/index.html') do
  it { should exist }
  its('type') { should cmp 'file' }
  it { should be_file }
  it { should_not be_directory }
  its('size') { should be > 256 }
  its('mode') { should cmp '00644' }
end

metadata.rb
name 'employee_database'
maintainer 'The Authors'
maintainer_email 'you@example.com'
license 'All Rights Reserved'
description 'Installs/Configures employee_database'
long_description 'Installs/Configures employee_database'
version '0.1.0'
chef_version '>= 13.0'

# The `issues_url` points to the location where issues for this cookbook are
# tracked.  A `View Issues` link will be displayed on this cookbook's page when
# uploaded to a Supermarket.
#
# issues_url 'https://github.com/<insert_org_here>/employee_database/issues'

# The `source_url` points to the development repository for this cookbook.  A
# `View Source` link will be displayed on this cookbook's page when uploaded to
# a Supermarket.
#
# source_url 'https://github.com/<insert_org_here>/employee_database'

depends 'apt', '~> 7.1.1'
depends 'nginx', '~> 9.0.0'
depends 'mysql', '~> 8.0'

recipes/database.rb
#
# Cookbook:: employee_database
# Recipe:: database
#
# Copyright:: 2019, The Authors, All Rights Reserved.

local_mysql_path = node['employee_database']['database']['mysql']['home_path']
local_mysql_user = data_bag_item('mysql', 'local_user')['user']
local_mysql_password = data_bag_item('mysql', 'local_user')['password']
test_db_path = node['employee_database']['database']['mysql']['database_path']
database_name = node['employee_database']['database']['mysql']['database_name']
remote_mysql_user = data_bag_item('mysql', 'remote_user')['user']
remote_mysql_pass = data_bag_item('mysql', 'remote_user')['password']

# Build cmd for connecting to mysql shell
init_connection=init_mysql_connection(database_name, local_mysql_user, local_mysql_password)

mysql_service database_name do
    port '3306'
    version '5.7'
    initial_root_password local_mysql_password
    action [:create, :start]
end

directory test_db_path do
    owner 'root'
    group 'root'
    mode '0644'
    recursive true
    action :create
end

git test_db_path do
    repository 'https://github.com/datacharmer/test_db.git'
    revision 'master'
    action :sync
end

directory local_mysql_path do
    owner 'mysql'
    group 'mysql'
    recursive true
    action :create
end

execute 'load test_db in mysql' do
    command <<-"EOF"
        cd #{test_db_path} && \
            #{init_connection} \
            < #{test_db_path}/employees.sql
    EOF
    user 'root'
    action :run
end

# Create remote user for accessing remotely
execute 'set remote mysql privileges' do
    command <<-"SQL"
        #{init_connection} -e "\
            CREATE USER '#{remote_mysql_user}'@'localhost' IDENTIFIED BY '#{remote_mysql_pass}';\
            GRANT ALL PRIVILEGES ON *.* TO '#{remote_mysql_user}'@'localhost' WITH GRANT OPTION;\
            CREATE USER '#{remote_mysql_user}'@'%' IDENTIFIED BY '#{remote_mysql_pass}';\
            GRANT ALL PRIVILEGES ON *.* TO '#{remote_mysql_user}'@'%' WITH GRANT OPTION;"
    SQL
    user 'root'
    action :run
end

recipes/app.rb
#
# Cookbook:: employee_database
# Recipe:: app
#
# Copyright:: 2019, The Authors, All Rights Reserved.

db_host_node=endpoint_search(node, 'role:database')
db_host=db_host_node[0].attributes['ipaddress']
remote_mysql_user=data_bag_item('mysql', 'remote_user')['user']
remote_mysql_password=data_bag_item('mysql', 'remote_user')['password']
query_output_file=node['employee_database']['database']['mysql']['query_output_file']
nginx_web_hosting_path='/usr/share/nginx/html/'

# Build cmd for connecting to mysql shell
init_remote_connection=init_remote_mysql_connection(db_host, remote_mysql_user, remote_mysql_password)

mysql_client_installation_package 'default' do
  version '5.7'
  action :create
end

execute 'set remote mysql privileges' do
  command <<-"SQL"
      #{init_remote_connection} -e "\
        use employees;\
        SELECT * FROM employees WHERE gender = 'M' AND birth_date = '1965-02-01' AND hire_date > '1990-01-01' ORDER BY last_name, first_name DESC;" \
        > "#{nginx_web_hosting_path}/#{query_output_file}"
SQL
  user 'root'
  action :run
end

template 'index.html' do
  path   "#{nginx_web_hosting_path}/index.html"
  source 'index.html.erb'
  mode '0644'
  variables({
    servername: node['hostname'],
    query_file: query_output_file
  })
end